### DISCUSSION TITLE
Go wrapper?

### DISCUSSION THREAD

**Discussion by legowerewolf** [zxing-cpp/zxing-cpp#752] [Ideas]
I saw that @siiky mentioned getting zxing-cpp working in Go through the C wrapper and cgo (https://github.com/zxing-cpp/zxing-cpp/discussions/551), so it *can* be done, but my coworker and I are having a rough time of it.

**Comment by axxel** [comment]
I have no experience with Go/cgo myself but in general, I would suggest you describe what issues you ran into to get some meaningful feedback.

**Comment by siiky** [comment]
Without any errors to look at, the short answer:

* `zxing-wrapper.cpp` with the functions you need for your usecase
* `zxing-wrapper.h` with the prototypes &c
* In the Go files using these functions:

```go
// #cgo CFLAGS: -Os -Wall -Wextra -Werror -fno-omit-frame-pointer
// #cgo CXXFLAGS: -Os -Wall -Wextra -Werror -fno-omit-frame-pointer
// #cgo LDFLAGS: -static -lm -lZXing
// #include "zxing-wrapper.h"
import "C"
```

Use the functions, constants, &c with `C.zxing_foo(...)`.

**Comment by legowerewolf** [comment]
Right, sorry. Just invited you two to a private repo with a basic demo program. It _works_, but it's not statically compiled. It needs `LD_LIBRARY_PATH` at runtime to find its libraries.

@siiky, do those flags produce a static binary?

**Reply by siiky** [reply]
I build ZXing statically, and use this `Makefile` rule to build the Go program:

```mk
ZXING_INSTALL_PREFIX :=

LDFLAGS := -L$(ZXING_INSTALL_PREFIX)/lib

CGO_CFLAGS := $(CFLAGS)
CGO_CXXFLAGS := $(CXXFLAGS)
CGO_LDFLAGS := $(LDFLAGS)

build:
	CGO_CFLAGS='$(CGO_CFLAGS) -I$(ZXING_INSTALL_PREFIX)/include' CGO_CXXFLAGS='$(CGO_CXXFLAGS) -I$(ZXING_INSTALL_PREFIX)/include' CGO_LDFLAGS='$(CGO_LDFLAGS)' go build -x -ldflags='-linkmode external -extldflags=-static'
```

```
$ ldd foo
	not a dynamic executable
$ file foo
foo: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=a0f780e593d6a59679138c6a261d75fc42edfb0b, for GNU/Linux 3.2.0, with debug_info, not stripped
```

**Reply by siiky** [reply]
I glanced at the repo. Looks overcomplicated to find the root cause. I would suggest starting from an MVP.

**Comment by axxel** [comment]
I forgot that this discussion has come up in the past. I just started a new one discussing a new effort that covers all the 3.0 features. See #1058. Closing this to focus the discussion there.

**Reply by axxel** [reply]
Here is a PR with a slightly updated wrapper code (comments welcome): https://github.com/zxing-cpp/zxing-cpp/pull/1064
